src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs

```diff
@@ -1144,97 +820,57 @@ private void SaveServerCredentialsToCache(Uri uri, uint authScheme, ICredentials
-                            retryRequest = true;
-                        }
-                        else if (lastError == (int)Interop.WinHttp.ERROR_WINHTTP_CLIENT_AUTH_CERT_NEEDED)
+                        if (state.RequestMessage.Content != null)
                         {
-                            // WinHttp will automatically drop any client SSL certificates that we
-                            // have pre-set into the request handle.  For security reasons, we don't
-                            // allow the certificate to be re-applied. But we need to tell WinHttp
-                            // that we don't have any certificate.
-                            SetNoClientCertificate(requestHandle);
-                            retryRequest = true;
+                            await InternalSendRequestBodyAsync(state, chunkedModeForSend).ConfigureAwait(false);
                         }
 
-                        else
+                        bool receivedResponse = await InternalReceiveResponseHeadersAsync(state);
+                        if (receivedResponse)
                         {
-                            throw WinHttpException.CreateExceptionUsingError(lastError);
+                            _authHelper.CheckResponseForAuthentication(
+                                state,
+                                ref proxyAuthScheme,
+                                ref serverAuthScheme);
                         }
-                    }
-                    else
-                    {
-                        ProcessResponse(
-                            state,
-                            requestHandle,
-                            ref proxyAuthScheme,
-                            ref serverAuthScheme,
-                            out retryRequest);
-                    }
-                } while (retryRequest);
-
-                // Clear callback function in WinHTTP once we have a final response
-                // and are ready to create the response message.
-                SetStatusCallback(requestHandle, null);
+                    } while (state.RetryRequest);
+                }
 
                 state.CancellationToken.ThrowIfCancellationRequested();
 
-                // Create HttpResponseMessage object.
-                responseMessage = CreateResponseMessage(requestHandle, state.RequestMessage);
+                responseMessage = WinHttpResponseParser.CreateResponseMessage(state, _doManualDecompressionCheck);
+
+                // Since the headers have been read, set the "receive" timeout to be based on each read
+                // call of the response body data. WINHTTP_OPTION_RECEIVE_TIMEOUT sets a timeout on each
+                // lower layer winsock read.
+                uint optionData = (uint)_receiveDataTimeout.TotalMilliseconds;
+                SetWinHttpOption(state.RequestHandle, Interop.WinHttp.WINHTTP_OPTION_RECEIVE_TIMEOUT, ref optionData);
             }
             catch (Exception ex)
```

Comment:
We don't re-throw from the `catch` block.  We save the exception and then set it into the TaskCompletionSource as appropriate.